### PR TITLE
APIv2: move all _autodetect_engine logic to the plugins

### PR DESCRIPTION
**Pull request by alexamici**
- [x] Relates to #4309 
 - [x] Tests updated
 - [x] Passes `isort . && black . && mypy . && flake8`
 - No user visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - No new functions/methods are listed in `api.rst`

**Changes affects only APIv2.**

This PR removes all remaining knowledge about what backend support from the generic code path in `apiv2.open_dataset`.

Once we make the switch to API v2 a big chunk of code can be removed from `api.py`.

### CODE REVIEW

File: xarray/backends/scipy_.py [hunk 1]
```diff
@@ -222,6 +223,29 @@ def close(self):
...
 
+def guess_can_open_scipy(store_spec):
+    # check byte header to determine file type
+    if isinstance(store_spec, bytes) or isinstance(store_spec, io.IOBase):
+        if isinstance(store_spec, bytes):
+            magic_number = store_spec[:8]
+        else:
+            if store_spec.tell() != 0:
+                raise ValueError(
+                    "file-like object read/write pointer not at zero "
```
Comment (keewis) [inline xarray/backends/scipy_.py:234]: is that a user-facing error? If so, it might be better to replace "zero" with "the start of the file.".

**Review by keewis — COMMENTED** [review]
I have a few minor comments, but otherwise this looks good to me

File: xarray/backends/plugins.py [hunk 2]
```diff
@@ -76,3 +78,21 @@ def list_engines():
...
+
+def guess_engine(store_spec):
+    engines = list_engines()
+
+    # use the pre-defined selection order for netCDF files
+    for engine in ["netcdf4", "h5netcdf", "scipy"]:
+        if engine in engines and engines[engine].guess_can_open(store_spec):
+            return engine
+
+    for engine, beckend in engines.items():
```
[the remark this answers is not in the document]
Comment (alexamici) [reply xarray/backends/plugins.py:91]: @keewis checks are not expensive, but I agree that this is ugly.

I think we need the concept of plugin *priority*, but that is more general than this PR. @aurghs I think we need to add a way to define priority.

I propose I'll add a comment and leave the redundant check for now.

File: xarray/backends/plugins.py [hunk 3]
```diff
@@ -76,3 +78,21 @@ def list_engines():
...
+        try:
+            if beckend.guess_can_open and beckend.guess_can_open(store_spec):
```
[the remark this answers is not in the document]
Comment (alexamici) [reply xarray/backends/plugins.py:93]: Thanks!

File: xarray/backends/scipy_.py (hunk 1 above)
Comment (alexamici) [reply xarray/backends/scipy_.py:234]: Yes, it is user facing an I like your suggestion.

I coped the message text from `_get_engine_from_magic_number`, but I can sync all the instances.

**Review by aurghs — APPROVED** [review]

**Pull request merged**